doc/specs/#888 - Com Api.md

```diff
@@ -0,0 +1,889 @@
+particularly confusing.
+
+# 5 API Details
+
+```c# (but really MIDL3)
+namespace Microsoft.Management.Deployment
+{
+    [contractversion(1)]
+    apicontract WindowsPackageManagerContract{};
+
+    /// State of the install.
+    [contract(Microsoft.Management.Deployment.WindowsPackageManagerContract, 1)]
+    enum AppInstallProgressState
+    {
+        /// The install is queued but not yet active. Cancellation of the IAsyncOperationWithProgress in this 
+        /// state will prevent the app from downloading or installing.
+        Queued,
+        /// The installer is downloading. Cancellation of the IAsyncOperationWithProgress in this state will 
+        /// end the download and prevent the app from installing.
+        Downloading,
+        /// The install is in progress. Cancellation of the IAsyncOperationWithProgress in this state will not
+        /// stop the installation or the post install cleanup.
+        Installing,
+        /// The installer has completed and cleanup actions are in progress. Cancellation of the 
+        /// IAsyncOperationWithProgress in this state will not stop cleanup or roll back the install.
+        PostInstall,
+        /// The operation has completed.
+        Finished,
+    };
+
+    /// Progress object for the install
+    /// DESIGN NOTE: percentage for the install as a whole is purposefully not included as there is no way to 
+    /// estimate progress when the installer is running.
+    [contract(Microsoft.Management.Deployment.WindowsPackageManagerContract, 1)]
+    struct InstallProgress
+    {
+        /// State of the install
+        AppInstallProgressState State;
+        /// DESIGN NOTE: BytesDownloaded may only be available for downloads done by Windows Package Manager itself.
+        /// Number of bytes downloaded if known
+        UInt64 BytesDownloaded;
+        /// DESIGN NOTE: BytesRequired may only be available for downloads done by Windows Package Manager itself.
+        /// Number of bytes required if known
+        UInt64 BytesRequired;
+        /// Download percentage completed
+        Single DownloadPercentage;
+        /// Install percentage if known.
+        Single InstallPercentage;
+    };
+
+    /// Result of the install
+    [contract(Microsoft.Management.Deployment.WindowsPackageManagerContract, 1)]
+    runtimeclass InstallResult
+    {
+        /// Used by a caller to correlate the install with a caller's data.
+        String CorrelationData{ get; };
+        /// Whether a restart is required to complete the install.
+        Boolean RebootRequired { get; };
+    }
+
+    /// IMPLEMENTATION NOTE: SourceOrigin from AppInstallerRepositorySource.h
+    /// Defines the origin of the app catalog details.
+    [contract(Microsoft.Management.Deployment.WindowsPackageManagerContract, 1)]
+    enum AppCatalogOrigin
+    {
+        /// Predefined means it came as part of the Windows Package Manager package and cannot be removed.
+        Predefined,

```

Comment:
Will it ever be possible that the origin is part of the Windows Package Manager and is can be removed?

Reply:
Not removed I don't think, but it may be possible for administrators to disable a predefined catalog so that it fails during Open.

Reply:
DISCUSSION: The package catalog could be disabled by various forms of policy. When we move to the CatalogReference, the OpenCatalog call returns an appropriate status like "disabledbypolicy"